### PR TITLE
Optimize locking on backfill history

### PR DESCRIPTION
**Pull request by yux0**
## What changed?
When backfill history via sync workflow state replication task, it acquires the current run id lock to make sure no concurrent backfill on history. Add the optimization to reduce the locking scope to the same history tree.

## Why?
We need this backfill history lock only for workflows share the same history tree. This only happens on reset workflows.

### CODE REVIEW

File: service/history/ndc/workflow_state_replicator.go [hunk 1]
```diff
@@ -279,29 +280,36 @@ func (r *WorkflowStateReplicatorImpl) backfillHistory(
 	namespaceID namespace.ID,
 	workflowID string,
 	runID string,
+	rootRunID string,
```
Comment (yycptt) [inline service/history/ndc/workflow_state_replicator.go:283]: nit: maybe rename to originalRunID here. rootRunID concept is used in other place for parent/child relationship.

**Review by xwduan — APPROVED** [review]

File: service/history/ndc/workflow_state_replicator.go [hunk 2]
```diff
@@ -226,6 +226,7 @@ func (r *WorkflowStateReplicatorImpl) SyncWorkflowState(
 		namespaceID,
 		wid,
 		rid,
+		branchInfo.GetTreeId(),
```
Comment (yycptt) [inline service/history/ndc/workflow_state_replicator.go:229]: Can you add some comments/TODO here and create an issue to track?

**Review by yycptt — APPROVED** [review]

File: service/history/ndc/workflow_state_replicator.go (hunk 1 above)
Comment (yux0) [reply service/history/ndc/workflow_state_replicator.go:283]: Will update the name

File: service/history/ndc/workflow_state_replicator.go (hunk 2 above)
Comment (yux0) [reply service/history/ndc/workflow_state_replicator.go:229]: Will do

**Pull request merged**